sdk/cxx/include/sawtooth/address_mapper.h

```diff
@@ -0,0 +1,95 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------*/
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "sawtooth/message_stream.h"
+
+namespace sawtooth {
+
+// Raised if any of the addresses generated violate
+// the merkel trie address constraints.
+class AddressFormatError: public std::runtime_error {
+ public:
+    explicit AddressFormatError(std::string const& error)
+        : std::runtime_error(error)
+    {}
+};
+
+
+// Helper class to provide mappings from domain identifiers
+// to merkle trie addresses. This class is designed to be overloaded, such
+// that implementers of TransactionProcessors can define their own mappings
+// from their native object ids to merkle trie addresses.
+//
+// The default implementation uses a SHA512 hash to map both the key and the
+// namespace to the mekle trie address. Collisions in the key mappings should
+// be uncommon but still possible, data storage behind the keys.
+//
+// Each instance of the mapper holds a namespace that is mapped to a
+// namespace prefix when it is constructed. If your transaction processor
+// operates on multiple namespaces, it is recommended that you use an instance
+// for each namespace.
+class AddressMapper {
+ public:
+    // Constructor for an address mapping object, Takes the
+    // unencoded namespace name, it will be mapping objects, to as a parameter.
+    // The namespace argument passed will be mapped to a namespace_prefix
+    explicit AddressMapper(const std::string& namespace_);
+    virtual ~AddressMapper() {}
+
+    AddressMapper (const AddressMapper&) = delete;
+    AddressMapper (const AddressMapper&&) = delete;
+    AddressMapper& operator= (const AddressMapper&) = delete;
+
+    // Maps an namespace string to an namespace address prefix.
+    // this provides a default implementation using SHA512.
+    // it is intended to be overriden by subclass implementations
+    // that can provide their own mapping of namspace strings to
+    // prefixes. The requirements are that the mapping needs to
+    // return a prefix that is 6 characters in length and consists
+    // of only lower case hexadecimal characters(0123456789abcdef).
+    virtual std::string MapNamespace(const std::string& key) const; 
+
+    // Maps an key string to a address, this provides a default
+    // implementation using SHA512.
+    // It is intended to be overriden by subclass implementations
+    // that can provide their own mappings. The requirements are that
+    // the mapping needs to return a string that is 128 characters
+    // in length and consists of only lower case hexadecimal characters.
+    virtual std::string MapKey(const std::string& key) const;
+
+
+    // returns the namespace prefix generated when the class was constructed.
+    std::string GetNamespacePrefix();
```

Comment:
Cannot be, since the virtual function overrides cannot be called in the constructor. GetNamespacePrefix generates and caches the namespace_prefix on first call.